# Examen de certificación – Intento 1

<strong>Objetivo:</strong> Este examen tiene como objetivo evaluar la capacidad del estudiante para limpiar, explorar, implementar y evaluar modelos de regresión en un dataset de precios de casas. Utilizaremos el «[House Prices Dataset](https://www.kaggle.com/competitions/house-prices-advanced-regression-techniques/data)» disponible en Kaggle. Los estudiantes deben demostrar habilidades prácticas en la manipulación de datos, creación de visualizaciones y modelado predictivo.

## Contexto y Descripción del Dataset

El «House Prices Dataset» contiene datos sobre diversas características de casas residenciales en Ames, Iowa. La tarea es predecir el precio final de cada casa (variable objetivo) en función de sus características.

## Diccionario de Datos


1. <strong>MSSubClass:</strong> Clase del edificio.
2. <strong>MSZoning:</strong> Clasificación de la zona.
3. <strong>LotFrontage:</strong> Frente del lote en pies lineales.
4. <strong>LotArea:</strong> Área del lote en pies cuadrados.
5. <strong>Street:</strong> Tipo de calle.
6. <strong>Alley:</strong> Tipo de callejón.
7. <strong>LotShape:</strong> Forma del lote.
8. <strong>LandContour:</strong> Contorno del terreno.
9. <strong>Utilities:</strong> Servicios públicos disponibles.
10. <strong>LotConfig:</strong> Configuración del lote.
11. <strong>LandSlope:</strong> Pendiente del terreno.
12. <strong>Neighborhood:</strong> Vecindario.
13. <strong>Condition1:</strong> Proximidad a diversas condiciones.
14. <strong>Condition2:</strong> Proximidad a diversas condiciones (segunda).
15. <strong>BldgType:</strong> Tipo de edificio.
16. <strong>HouseStyle:</strong> Estilo de la casa.
17. <strong>OverallQual:</strong> Calidad general del material y acabado.
18. <strong>OverallCond:</strong> Condición general del edificio.
19. <strong>YearBuilt:</strong> Año de construcción original.
20. <strong>YearRemodAdd:</strong> Año de remodelación o adición.
21. <strong>RoofStyle:</strong> Estilo del techo.
22. <strong>RoofMatl:</strong> Material del techo.
23. <strong>Exterior1st:</strong> Revestimiento exterior de la casa.
24. <strong>Exterior2nd:</strong> Revestimiento exterior de la casa (segundo).
25. <strong>MasVnrType:</strong> Tipo de revestimiento de mampostería.
26. <strong>MasVnrArea:</strong> Área de revestimiento de mampostería en pies cuadrados.
27. <strong>ExterQual:</strong> Calidad del material exterior.
28. <strong>ExterCond:</strong> Condición del material exterior.
29. <strong>Foundation:</strong> Tipo de fundación.
30. <strong>BsmtQual:</strong> Altura del sótano.
31. <strong>BsmtCond:</strong> Condición general del sótano.
32. <strong>BsmtExposure:</strong> Exposición del sótano.
33. <strong>BsmtFinType1:</strong> Calidad del acabado del sótano.
34. <strong>BsmtFinSF1:</strong> Área terminada del sótano en pies cuadrados.
35. <strong>BsmtFinType2:</strong> Calidad del acabado del sótano (segunda).
36. <strong>BsmtFinSF2:</strong> Área terminada del sótano (segunda) en pies cuadrados.
37. <strong>BsmtUnfSF:</strong> Área no terminada del sótano en pies cuadrados.
38. <strong>TotalBsmtSF:</strong> Área total del sótano en pies cuadrados.
39. <strong>Heating:</strong> Tipo de calefacción.
40. <strong>HeatingQC:</strong> Calidad y condición de la calefacción.
41. <strong>CentralAir:</strong> Aire acondicionado central.
42. <strong>Electrical:</strong> Sistema eléctrico.
43. <strong>1stFlrSF:</strong> Área del primer piso en pies cuadrados.
44. <strong>2ndFlrSF:</strong> Área del segundo piso en pies cuadrados.
45. <strong>LowQualFinSF:</strong> Área terminada de baja calidad en pies cuadrados.
46. <strong>GrLivArea:</strong> Área habitable por encima del nivel del suelo en pies cuadrados.
47. <strong>BsmtFullBath:</strong> Número de baños completos en el sótano.
48. <strong>BsmtHalfBath:</strong> Número de medios baños en el sótano.
49. <strong>FullBath:</strong> Número de baños completos por encima del nivel del suelo.
50. <strong>HalfBath:</strong> Número de medios baños por encima del nivel del suelo.
51. <strong>Bedroom:</strong> Número de dormitorios por encima del nivel del suelo.
52. <strong>Kitchen:</strong> Número de cocinas.
53. <strong>KitchenQual:</strong> Calidad de la cocina.
54. <strong>TotRmsAbvGrd:</strong> Número total de habitaciones por encima del nivel del suelo (excluyendo baños).
55. <strong>Functional:</strong> Funcionalidad del hogar.
56. <strong>Fireplaces:</strong> Número de chimeneas.
57. <strong>FireplaceQu:</strong> Calidad de la chimenea.
58. <strong>GarageType:</strong> Ubicación del garaje.
59. <strong>GarageYrBlt:</strong> Año en que se construyó el garaje.
60. <strong>GarageFinish:</strong> Interior terminado del garaje.
61. <strong>GarageCars:</strong> Tamaño del garaje en capacidad de coches.
62. <strong>GarageArea:</strong> Área del garaje en pies cuadrados.
63. <strong>GarageQual:</strong> Calidad del garaje.
64. <strong>GarageCond:</strong> Condición del garaje.
65. <strong>PavedDrive:</strong> Entrada pavimentada.
66. <strong>WoodDeckSF:</strong> Área de la terraza de madera en pies cuadrados.
67. <strong>OpenPorchSF:</strong> Área del porche abierto en pies cuadrados.
68. <strong>EnclosedPorch:</strong> Área del porche cerrado en pies cuadrados.
69. <strong>3SsnPorch:</strong> Área del porche de tres estaciones en pies cuadrados.
70. <strong>ScreenPorch:</strong> Área del porche de la pantalla en pies cuadrados.
71. <strong>PoolArea:</strong> Área de la piscina en pies cuadrados.
72. <strong>PoolQC:</strong> Calidad de la piscina.
73. <strong>Fence:</strong> Calidad de la cerca.
74. <strong>MiscFeature:</strong> Característica miscelánea no cubierta en otras categorías.
75. <strong>MiscVal:</strong> Valor misceláneo.
76. <strong>MoSold:</strong> Mes en que se vendió la propiedad.
77. <strong>YrSold:</strong> Año en que se vendió la propiedad.
78. <strong>SaleType:</strong> Tipo de venta.
79. <strong>SaleCondition:</strong> Condición de venta.
80. <strong>SalePrice:</strong> Precio de venta (variable objetivo).

## Requisitos

### 1. Limpieza de Datos:

* <strong>Identificación y eliminación de valores duplicados:</strong> Asegúrate de que no haya registros duplicados que puedan sesgar los resultados del análisis.
* <strong>Verificación y ajuste de tipos de datos:</strong> Verifica que cada columna tenga el tipo de dato correcto (numérico o categórico) y ajusta si es necesario.
* <strong>Corrección de inconsistencias en valores categóricos:</strong> Revisa las categorías de las variables y unifica aquellos valores que puedan estar escritos de diferentes maneras pero que representen lo mismo.
* <strong>Manejo de valores faltantes adecuadamente:</strong> Identifica y maneja los valores faltantes utilizando técnicas apropiadas como la imputación de la mediana, media o moda, según corresponda.

### 2. Exploración de Datos:

* <strong>Visualizaciones univariadas y multivariadas:</strong> Crea histogramas, gráficos de barras, diagramas de dispersión y mapas de calor para entender la distribución y las relaciones entre las variables.
* <strong>Estadísticas descriptivas:</strong> Calcula medidas de tendencia central (media, mediana, moda) y de dispersión (rango, desviación estándar) para cada característica del dataset.

### 3. Implementación de Modelos:

* <strong>Modelos de Regresión:</strong> Implementa modelos de Linear Regression y LightGBM (LGBM).
* <strong>Evaluación de Modelos:</strong> Evalúa los modelos utilizando métricas como MSE, RMSE, y R^2.
* <strong>Comparación de Rendimiento:</strong> Compara los resultados de ambos modelos y discute cuál es el más adecuado para este dataset.

## Entrega

Los estudiantes deben entregar un archivo .ipynb comentado que incluya:

* Proceso completo de limpieza y preprocesamiento de datos.
* Visualizaciones y estadísticas descriptivas.
* Implementación y evaluación de los modelos de regresión.
* Análisis comparativo del rendimiento de los modelos.

Además, el archivo debe subirse a GitHub con un tag de liberación (release tag) que permita identificar la entrega final.

## Consideraciones Éticas y Tecnológicas

### Consideraciones Éticas:

* <strong>Transparencia y Reproducibilidad:</strong> Asegúrate de que todos los pasos del análisis sean claros y reproducibles. Otros investigadores deben poder seguir tus pasos y llegar a los mismos resultados.
* <strong>Imparcialidad y Sesgo:</strong> Revisa si existen sesgos en los datos que puedan afectar la imparcialidad del modelo. Es importante que los modelos no discriminen injustamente entre diferentes grupos de datos.

### Consideraciones Tecnológicas:

* <strong>Herramientas Utilizadas:</strong> Utiliza herramientas estándar como Python, Jupyter Notebook, Pandas, Scikit-learn, Matplotlib y Seaborn.
* <strong>Escalabilidad:</strong> Considera cómo las técnicas aplicadas podrían escalarse para manejar conjuntos de datos más grandes y complejos.
* <strong>Optimización de Modelos:</strong> Aunque este examen no se enfoca en la optimización de hiperparámetros, se debe tener en cuenta para futuras implementaciones y mejorar el rendimiento de los modelos.

# Resolución

## Carga de datos

Se cargan primero los datos de entrenamiento

In [526]:
import pandas as pd

camino_ds_train = "../data/train.csv"
# Usamos el parametro keep_default_na porque existen
na_values = [
    "", 
    "#N/A", 
    "#N/A N/A", 
    "#NA", 
    "-1.#IND", 
    "-1.#QNAN", 
    "-NaN", 
    "-nan", 
    "1.#IND", 
    "1.#QNAN", 
    "<NA>", 
    "N/A", 
    # "NA", 
    "NULL", 
    "NaN", 
    "n/a", 
    "nan", 
    "null",
]

# Se usan los parámetros na_values y keep_default_valures de acuerdo
# a lo especificado aquí: https://stackoverflow.com/a/66715400
# Esto se realiza porque existen valores NA en el dataset que son reconocidos como nulos.
# Investigando en data_description.txt también vemos que estos valores son efectivamente válidos para
# ciertas columnas.
df_train = pd.read_csv(camino_ds_train, delimiter=",", na_values=na_values, keep_default_na=False)

df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Id             1460 non-null   int64 
 1   MSSubClass     1460 non-null   int64 
 2   MSZoning       1460 non-null   object
 3   LotFrontage    1460 non-null   object
 4   LotArea        1460 non-null   int64 
 5   Street         1460 non-null   object
 6   Alley          1460 non-null   object
 7   LotShape       1460 non-null   object
 8   LandContour    1460 non-null   object
 9   Utilities      1460 non-null   object
 10  LotConfig      1460 non-null   object
 11  LandSlope      1460 non-null   object
 12  Neighborhood   1460 non-null   object
 13  Condition1     1460 non-null   object
 14  Condition2     1460 non-null   object
 15  BldgType       1460 non-null   object
 16  HouseStyle     1460 non-null   object
 17  OverallQual    1460 non-null   int64 
 18  OverallCond    1460 non-null

Encontramos las 80 columnas especificadas en el diccionario de datos y una columna de Id. También notamos que existen 1460 filas en el dataset de entrenamiento.

In [527]:
df_train.head(10)

Id  MSSubClass MSZoning LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL          65     8450   Pave    NA      Reg   
1   2          20       RL          80     9600   Pave    NA      Reg   
2   3          60       RL          68    11250   Pave    NA      IR1   
3   4          70       RL          60     9550   Pave    NA      IR1   
4   5          60       RL          84    14260   Pave    NA      IR1   
5   6          50       RL          85    14115   Pave    NA      IR1   
6   7          20       RL          75    10084   Pave    NA      Reg   
7   8          60       RL          NA    10382   Pave    NA      IR1   
8   9          50       RM          51     6120   Pave    NA      Reg   
9  10         190       RL          50     7420   Pave    NA      Reg   

  LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub  ...        0     NA     NA          NA       0   
1         Lvl    AllPub  ...        0     NA     NA          NA       0   
2         Lvl    AllPub  ...        0     NA     NA          NA       0   
3         Lvl    AllPub  ...        0     NA     NA          NA       0   
4         Lvl    AllPub  ...        0     NA     NA          NA       0   
5         Lvl    AllPub  ...        0     NA  MnPrv        Shed     700   
6         Lvl    AllPub  ...        0     NA     NA          NA       0   
7         Lvl    AllPub  ...        0     NA     NA        Shed     350   
8         Lvl    AllPub  ...        0     NA     NA          NA       0   
9         Lvl    AllPub  ...        0     NA     NA          NA       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  
5     10   2009        WD         Normal     143000  
6      8   2007        WD         Normal     307000  
7     11   2009        WD         Normal     200000  
8      4   2008        WD        Abnorml     129900  
9      1   2008        WD         Normal     118000  

[10 rows x 81 columns]

Se cargan los datos de prueba

In [528]:
camino_ds_test = "../data/test.csv"
df_test = pd.read_csv(camino_ds_train, delimiter=",", na_values=na_values, keep_default_na=False)

df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Id             1460 non-null   int64 
 1   MSSubClass     1460 non-null   int64 
 2   MSZoning       1460 non-null   object
 3   LotFrontage    1460 non-null   object
 4   LotArea        1460 non-null   int64 
 5   Street         1460 non-null   object
 6   Alley          1460 non-null   object
 7   LotShape       1460 non-null   object
 8   LandContour    1460 non-null   object
 9   Utilities      1460 non-null   object
 10  LotConfig      1460 non-null   object
 11  LandSlope      1460 non-null   object
 12  Neighborhood   1460 non-null   object
 13  Condition1     1460 non-null   object
 14  Condition2     1460 non-null   object
 15  BldgType       1460 non-null   object
 16  HouseStyle     1460 non-null   object
 17  OverallQual    1460 non-null   int64 
 18  OverallCond    1460 non-null

En este caso notamos que solamente existen 79 columnas mas la columna de Id. Verificamos si esto se debe a que la variable a predecir no está en este dataset o si es otra variable la que falta.

In [529]:
columnas_df_train = set(df_train.columns.tolist())
columnas_df_test = set(df_test.columns.tolist())

diferencia = columnas_df_train.difference(columnas_df_test)
print(f"Cant. de columnas únicas en train: {len(list(columnas_df_train))}")
print(f"Cant. de columnas únicas en test: {len(list(columnas_df_test))}")
print(f"Las columnas diferentes son: {diferencia}")

Cant. de columnas únicas en train: 81
Cant. de columnas únicas en test: 81
Las columnas diferentes son: set()


Efectivamente, la única columna que no existe en test es la de SalePrice, que es la columna a ser predicha.

Adicionalmente, para evitar sesgos en el modelo de predicción, sólamente se realizará la limpieza y exploración del dataset de entrenamiento. Esto se hace con el objetivo de no intentar optimizar el modelo viendo datos del dataset de pruebas. Si se analizará al momento de realizar las predicciones, pero no se incluirán en ninguna parte del proceso previo al entrenamiento.

In [530]:
df = df_train.copy()
del df_train
del df_test

In [531]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Id             1460 non-null   int64 
 1   MSSubClass     1460 non-null   int64 
 2   MSZoning       1460 non-null   object
 3   LotFrontage    1460 non-null   object
 4   LotArea        1460 non-null   int64 
 5   Street         1460 non-null   object
 6   Alley          1460 non-null   object
 7   LotShape       1460 non-null   object
 8   LandContour    1460 non-null   object
 9   Utilities      1460 non-null   object
 10  LotConfig      1460 non-null   object
 11  LandSlope      1460 non-null   object
 12  Neighborhood   1460 non-null   object
 13  Condition1     1460 non-null   object
 14  Condition2     1460 non-null   object
 15  BldgType       1460 non-null   object
 16  HouseStyle     1460 non-null   object
 17  OverallQual    1460 non-null   int64 
 18  OverallCond    1460 non-null

## 1. Limpieza de datos

### Identificación y eliminación de valores duplicados

In [532]:
duplicados = df.duplicated()

cant_registros_duplicados = duplicados.sum()

print(f"Cant. de Registros Duplicados general: {cant_registros_duplicados}")

Cant. de Registros Duplicados general: 0


En general no parecen haber registros duplicados. Verificamos ahora por la columna Id

In [533]:
duplicados_id = df.duplicated(["Id"])

cant_registros_duplicados_id = duplicados_id.sum()

print(f"Cant. de Registros Duplicados Id: {cant_registros_duplicados_id}")

Cant. de Registros Duplicados Id: 0


No parecen existir IDs duplicadas, por lo que se puede considerar que no hay duplicados en el dataset de train. Adicionalmente, la columna Id posiblemente venga de algún tipo de dato relacional, y en general no servirá para realizar predicciones, por lo que se opta por eliminar esta columna. Este paso también deberá realizarse en el dataset de pruebas una vez que el modelo esté entrenado.

In [534]:
df.drop("Id", axis=1, inplace=True)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   MSSubClass     1460 non-null   int64 
 1   MSZoning       1460 non-null   object
 2   LotFrontage    1460 non-null   object
 3   LotArea        1460 non-null   int64 
 4   Street         1460 non-null   object
 5   Alley          1460 non-null   object
 6   LotShape       1460 non-null   object
 7   LandContour    1460 non-null   object
 8   Utilities      1460 non-null   object
 9   LotConfig      1460 non-null   object
 10  LandSlope      1460 non-null   object
 11  Neighborhood   1460 non-null   object
 12  Condition1     1460 non-null   object
 13  Condition2     1460 non-null   object
 14  BldgType       1460 non-null   object
 15  HouseStyle     1460 non-null   object
 16  OverallQual    1460 non-null   int64 
 17  OverallCond    1460 non-null   int64 
 18  YearBuilt      1460 non-null

### Verificación y ajuste de tipos de datos

Empezamos por buscar variables de tipo booleanas y convertirlas si es que existen.

In [535]:
columnas = df.columns.tolist()

columnas_convertidas_booleanas = []

valores_columnas_booleanas_potenciales = [
    [0, 1],
    [0.0, 1.0],
    ["false", "true"],
    ["no", "yes"],
    ["no", "si"],
    ["N", "Y"],
]


def _convertir_valor_a_booleano(x):
    if x in [0, 0.0, "false", "no", "no", "N"]:
        return False

    if x in [1, 1.0, "true", "yes", "si", "Y"]:
        return True

    return x


for columna in columnas:
    valores_unicos = df[columna].dropna().unique()

    columna_booleana_encontrada = False

    if len(valores_unicos) == 2:
        print(f"Valores de columna '{columna}': {str(valores_unicos)}")

        for posibilidad_valores_booleanos in valores_columnas_booleanas_potenciales:
            if set(valores_unicos) == set(posibilidad_valores_booleanos):
                print(f"La columna '{columna}' es booleana. Será convertida.")
                df[columna] = df[columna].apply(_convertir_valor_a_booleano)
                df[columna] = df[columna].astype("bool")

                columna_booleana_encontrada = True
                break

    if columna_booleana_encontrada:
        columnas_convertidas_booleanas.append(columna)


print(f"Columnas convertidas a booleanas: {str(columnas_convertidas_booleanas)}")

Valores de columna 'Street': ['Pave' 'Grvl']
Valores de columna 'Utilities': ['AllPub' 'NoSeWa']
Valores de columna 'CentralAir': ['Y' 'N']
La columna 'CentralAir' es booleana. Será convertida.
Columnas convertidas a booleanas: ['CentralAir']


Se ha encontrado sólamente una columna potencialmente booleana y se convirtió usando la lógica de arriba. No hace falta preocuparse por valores nulos en esta parte ya que se retornan como están si es que no se encuentran en la lista de tipos booleanos potenciales.

In [536]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   MSSubClass     1460 non-null   int64 
 1   MSZoning       1460 non-null   object
 2   LotFrontage    1460 non-null   object
 3   LotArea        1460 non-null   int64 
 4   Street         1460 non-null   object
 5   Alley          1460 non-null   object
 6   LotShape       1460 non-null   object
 7   LandContour    1460 non-null   object
 8   Utilities      1460 non-null   object
 9   LotConfig      1460 non-null   object
 10  LandSlope      1460 non-null   object
 11  Neighborhood   1460 non-null   object
 12  Condition1     1460 non-null   object
 13  Condition2     1460 non-null   object
 14  BldgType       1460 non-null   object
 15  HouseStyle     1460 non-null   object
 16  OverallQual    1460 non-null   int64 
 17  OverallCond    1460 non-null   int64 
 18  YearBuilt      1460 non-null

Luego, verificamos las columnas de tipo objeto y cadena para ver los únicos por columna.

In [537]:
columnas_objeto_cadena = df.select_dtypes(include=["object", "string"]).columns.tolist()


for columna in columnas_objeto_cadena:
    valores_unicos = df[columna].dropna().unique().tolist()
    print(f"Valores únicos para columna '{columna}': {str(valores_unicos)}")

Valores únicos para columna 'MSZoning': ['RL', 'RM', 'C (all)', 'FV', 'RH']
Valores únicos para columna 'LotFrontage': ['65', '80', '68', '60', '84', '85', '75', 'NA', '51', '50', '70', '91', '72', '66', '101', '57', '44', '110', '98', '47', '108', '112', '74', '115', '61', '48', '33', '52', '100', '24', '89', '63', '76', '81', '95', '69', '21', '32', '78', '121', '122', '40', '105', '73', '77', '64', '94', '34', '90', '55', '88', '82', '71', '120', '107', '92', '134', '62', '86', '141', '97', '54', '41', '79', '174', '99', '67', '83', '43', '103', '93', '30', '129', '140', '35', '37', '118', '87', '116', '150', '111', '49', '96', '59', '36', '56', '102', '58', '38', '109', '130', '53', '137', '45', '106', '104', '42', '39', '144', '114', '128', '149', '313', '168', '182', '138', '160', '152', '124', '153', '46']
Valores únicos para columna 'Street': ['Pave', 'Grvl']
Valores únicos para columna 'Alley': ['NA', 'Grvl', 'Pave']
Valores únicos para columna 'LotShape': ['Reg', 'IR1', 'IR2'

Procedemos a investigar por qué se da esto.

In [538]:
columnas_a_investigar = ["LotFrontage", "MasVnrArea", "GarageYrBlt"]

for columna in columnas_a_investigar:
    valores_no_nros = []

    for el in df.loc[:, columna]:
        try:
            int(el)
        except ValueError:
            valores_no_nros.append(el)

    valores_no_nros = list(set(valores_no_nros))
    print(f"Valores no numéricos para columna '{columna}': {str(valores_no_nros)}")

Valores no numéricos para columna 'LotFrontage': ['NA']
Valores no numéricos para columna 'MasVnrArea': ['NA']
Valores no numéricos para columna 'GarageYrBlt': ['NA']


Según lo visto los valores NA en este caso indican vacíos. Estos serán marcados como nulos en las variables numéricas.

In [539]:
def _convertir_NA_a_Nan(x):
    if x == "NA":
        return pd.NA

    return int(x)


for columna in columnas_a_investigar:
    # Int64 es el valor usado por pandas para identificar enteros
    # notar que si se usa int64 se da error porque existen valores nulos.
    # Esto se manejará al momento de imputar
    df[columna] = df[columna].apply(_convertir_NA_a_Nan).astype("Int64")

In [540]:
df[[*columnas_a_investigar]].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   LotFrontage  1201 non-null   Int64
 1   MasVnrArea   1452 non-null   Int64
 2   GarageYrBlt  1379 non-null   Int64
dtypes: Int64(3)
memory usage: 38.6 KB


Al parecer, todas las columnas de tipo objeto y cadena son categóricas. No se encuentra ninguna columna con valores extraños.

Se procede a limpiar las cadenas.

In [541]:
from utils.eda_utils import limpiar_cadena

columnas_objeto_cadena = df.select_dtypes(include=["object", "string"]).columns.tolist()


for columna in columnas_objeto_cadena:
    df[columna] = df[columna].apply(limpiar_cadena)

Verificamos de vuelta los valores únicos para corroborar que no existan errores de conversión.

In [542]:
for columna in columnas_objeto_cadena:
    valores_unicos = df[columna].unique().tolist()
    print(f"Valores únicos para columna '{columna}': {str(valores_unicos)}")

Valores únicos para columna 'MSZoning': ['rl', 'rm', 'c (all)', 'fv', 'rh']
Valores únicos para columna 'Street': ['pave', 'grvl']
Valores únicos para columna 'Alley': ['na', 'grvl', 'pave']
Valores únicos para columna 'LotShape': ['reg', 'ir1', 'ir2', 'ir3']
Valores únicos para columna 'LandContour': ['lvl', 'bnk', 'low', 'hls']
Valores únicos para columna 'Utilities': ['allpub', 'nosewa']
Valores únicos para columna 'LotConfig': ['inside', 'fr2', 'corner', 'culdsac', 'fr3']
Valores únicos para columna 'LandSlope': ['gtl', 'mod', 'sev']
Valores únicos para columna 'Neighborhood': ['collgcr', 'veenker', 'crawfor', 'noridge', 'mitchel', 'somerst', 'nwames', 'oldtown', 'brkside', 'sawyer', 'nridght', 'names', 'sawyerw', 'idotrr', 'meadowv', 'edwards', 'timber', 'gilbert', 'stonebr', 'clearcr', 'npkvill', 'blmngtn', 'brdale', 'swisu', 'blueste']
Valores únicos para columna 'Condition1': ['norm', 'feedr', 'posn', 'artery', 'rrae', 'rrnn', 'rran', 'posa', 'rrne']
Valores únicos para columna

Ya parecen estar normalizadas las columnas. También, se nota que potencialmente existen varias columnas que son potencialmente ordinales. Por ej: BsmtQual, GarageQual. Como son calidades, mientras más alto sea mejor. Por lo cual, se debe realizar una codificación teniendo en cuenta esto.

Convertimos las columnas a categóricas.

In [ ]:
# Se elige el 5% como valor para empezar a considerar una columna como categórica.
# Si se supera esta cantidad la columna se dejará como cadena.
porcentaje_max_columnas_categoricas = 0.05
cant_filas_df = df.shape[0]

columnas_a_excluir = []
columnas_convertidas_a_categoricas = []
columnas_convertidas_a_string = []

for columna in columnas_objeto_cadena:
    valores_unicos = df[columna].dropna().unique().tolist()
    print(f"Cant. únicos para columna '{columna}': {len(valores_unicos)}")
    if len(valores_unicos) <= (porcentaje_max_columnas_categoricas * cant_filas_df):
        
        print(f"Valores únicos columna '{columna}': {str(valores_unicos)}")
        print(f"Se procede a convertir la columna '{columna}' a categórica")
        df[columna] = df[columna].astype("category")
        columnas_convertidas_a_categoricas.append(columna)
    else:
        print(f"No se convierte la columna '{columna}' a categórica, se opta por dejarla como string")
        df[columna] = df[columna].astype("string")
        columnas_convertidas_a_string

    del valores_unicos

print("\n")
print(f"Cant. de columnas analizadas: {len(columnas_objeto_cadena)}")
print(f"Cant. de columnas convertidas a category: {len(columnas_convertidas_a_categoricas)}")
print(f"Cant. de columnas convertidas a string: {len(columnas_convertidas_a_string)}")
print(f"Columnas convertidas a categóricas: {str(columnas_convertidas_a_categoricas)}")
print(f"Columnas convertidas a string: {str(columnas_convertidas_a_string)}")

Cant. únicos para columna 'MSZoning': 5
Valores únicos columna 'MSZoning': ['rl', 'rm', 'c (all)', 'fv', 'rh']
Se procede a convertir la columna 'MSZoning' a categórica
Cant. únicos para columna 'Street': 2
Valores únicos columna 'Street': ['pave', 'grvl']
Se procede a convertir la columna 'Street' a categórica
Cant. únicos para columna 'Alley': 3
Valores únicos columna 'Alley': ['na', 'grvl', 'pave']
Se procede a convertir la columna 'Alley' a categórica
Cant. únicos para columna 'LotShape': 4
Valores únicos columna 'LotShape': ['reg', 'ir1', 'ir2', 'ir3']
Se procede a convertir la columna 'LotShape' a categórica
Cant. únicos para columna 'LandContour': 4
Valores únicos columna 'LandContour': ['lvl', 'bnk', 'low', 'hls']
Se procede a convertir la columna 'LandContour' a categórica
Cant. únicos para columna 'Utilities': 2
Valores únicos columna 'Utilities': ['allpub', 'nosewa']
Se procede a convertir la columna 'Utilities' a categórica
Cant. únicos para columna 'LotConfig': 5
Valores ú

Según las estadísticas, todas las columnas de tipo objeto fueron convertidas a category.

In [544]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype   
---  ------         --------------  -----   
 0   MSSubClass     1460 non-null   int64   
 1   MSZoning       1460 non-null   category
 2   LotFrontage    1201 non-null   Int64   
 3   LotArea        1460 non-null   int64   
 4   Street         1460 non-null   category
 5   Alley          1460 non-null   category
 6   LotShape       1460 non-null   category
 7   LandContour    1460 non-null   category
 8   Utilities      1460 non-null   category
 9   LotConfig      1460 non-null   category
 10  LandSlope      1460 non-null   category
 11  Neighborhood   1460 non-null   category
 12  Condition1     1460 non-null   category
 13  Condition2     1460 non-null   category
 14  BldgType       1460 non-null   category
 15  HouseStyle     1460 non-null   category
 16  OverallQual    1460 non-null   int64   
 17  OverallCond    1460 non-null   in

Dejamos también anotadas las columnas categóricas que en el diccionario de datos son de calificaciones, o ordinales. La descripción puede encontrarse [aquí](../data/data_description.txt).

In [546]:
columnas_numericas_categoricas_ordinales = [
    "OverallQual",
    "OverallCond"
]

columnas_cadenas_categoricas_ordinales = [
    "ExterQual"
    "ExterCond",
    "BsmtQual",
    "BsmtCond",
    "BsmtExposure",
    "BsmtFinType1",
    "BsmtFinType2"
    "HeatingQC",
    "KitchenQual",
    "Functional",
    "FireplaceQu",
    "GarageFinish",
    "GarageQual",
    "GarageCond",
    "PoolQC",
    "Fence",
]

Como las columnas de tipo cadena ya están todas convertidas a category, convertimos también las columnas numéricas a category. Esto se hace ya que las columnas son efectivamente categóricas (tienen un rango del 1 al 10 según data_description.txt). Estos rangos se verificarán que estén correctos al evaluar las estadísticas descriptivas.

In [547]:
for columna in columnas_numericas_categoricas_ordinales:
    df[columna] = df[columna].astype("category")

Vemos el resultado final de la limpieza.

In [548]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype   
---  ------         --------------  -----   
 0   MSSubClass     1460 non-null   int64   
 1   MSZoning       1460 non-null   category
 2   LotFrontage    1201 non-null   Int64   
 3   LotArea        1460 non-null   int64   
 4   Street         1460 non-null   category
 5   Alley          1460 non-null   category
 6   LotShape       1460 non-null   category
 7   LandContour    1460 non-null   category
 8   Utilities      1460 non-null   category
 9   LotConfig      1460 non-null   category
 10  LandSlope      1460 non-null   category
 11  Neighborhood   1460 non-null   category
 12  Condition1     1460 non-null   category
 13  Condition2     1460 non-null   category
 14  BldgType       1460 non-null   category
 15  HouseStyle     1460 non-null   category
 16  OverallQual    1460 non-null   category
 17  OverallCond    1460 non-null   ca

Notar que los valores Int64 son la version de pandas para reconocer valores numéricos, se usa la notación en mayúscula para diferenciar del tipo de numpy int64 que no incluye valores nulos.